HenryJia: auto-move data decorator

awaelchli opened the pull request:
## What does this PR do?
Continuation of the work of @HenryJia from PR #1526, fixes #1412
This is currently based on PR #1756.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

Borda commented:
@awaelchli seems this is unblocked now, mind resolve conflicts? :rabbit: 
@HenryJia mind check the update?

awaelchli commented:
Thanks for checking @HenryJia . I rebased and tests run except for one. 
I added a doctest but it requires the GPU. It fails if the machine only has CPU. 
I'm looking into that, I would really like to have that doctest. But otherwise PR is ready.

Borda commented:
> I'm looking into that, I would really like to have that doctest. But otherwise PR is ready.

I would like to have the doctest for it too, but maybe it would be easier to have the particular code just as an example and get this merge and later try to find a way to have it as doctest...

Borda commented on tests/core/test_decorators.py:
```diff
@@ -0,0 +1,33 @@
+def test_auto_move_data(src_device, dest_device):
+    """ Test that the decorator moves the data to the device the model is on. """
+
+    class CurrentModel(EvalModelTemplate):
+        pass
+
+    # apply the decorator
+    CurrentModel.forward = auto_move_data(CurrentModel.forward)
+
+    model = CurrentModel()
```
```suggestion
    model = EvalModelTemplate()
    # apply the decorator
    model.forward = auto_move_data(model.forward)
```
why pure `EvalModelTemplate` is not enough and you need a child?

awaelchli replied:
unfortunately not working. the decorator needs to be applied to the class definition. 
So we can do `EvalModelTemplate.forward = auto_move_data(EvalModelTemplate.forward)` but then it is applied for all following tests too.

Borda replied:
yeah, that looks easier :rabbit:

awaelchli replied:
this has side effects on other tests (and are failing). the decorator should only be applied at test time, when we e.g. only have one gpu. 
I think it's cleaner if we applied the decorator only locally in this test.

Borda commented:
@HenryJia @awaelchli could we get this done? :raccoon:

awaelchli commented:
@Borda If the decorator is enough, I think it is done module review. However this is probably not going to be merged very soon since there are these huge PRs on the top of the list.

awaelchli commented:
@SkafteNicki thanks, good suggestion. I added a short description to the deployment section.

awaelchli commented:
Hey @HenryJia we made it!! good work ✋

HenryJia commented:
Awesome cheers <3 thanks for continuing it when I was too busy
